cql3/statements/create_keyspace_statement.cc

```diff
@@ -148,34 +148,37 @@ std::optional<sstring> check_restricted_replication_strategy(
     if (!attrs.get_replication_strategy_class()) {
         return std::nullopt;
     }
+
+    static const std::unordered_map<sstring, locator::replication_strategy_type> rs_to_type = {
+            {"org.apache.cassandra.locator.SimpleStrategy", locator::replication_strategy_type::simple},
+            {"org.apache.cassandra.locator.NetworkTopologyStrategy", locator::replication_strategy_type::network_topology},
+            {"org.apache.cassandra.locator.EverywhereStrategy", locator::replication_strategy_type::everywhere_topology},
+            {"org.apache.cassandra.locator.LocalStrategy", locator::replication_strategy_type::local}
+    };
```

Comment:
This feels like it must be duplicating information that is available elsewhere, I'm not sure where. Maybe some sort of registrator of strategies (e.g., I see locator/simple_strategy.cc registers org.apache.cassandra.locator.SimpleStrategy"). Could we find it elsewhere?

But even more fundamentally, I don't see why this map is even needed: Since the name "SimpleStrategy" (for example) is the same one used on both CREATE statement and in the configuration, why can't we just compare names? Why do we need to go through translation to the locator::replication_strategy_type enum?

Reply:
> This feels like it must be duplicating information that is available elsewhere, I'm not sure where. Maybe some sort of registrator of strategies (e.g., I see locator/simple_strategy.cc registers org.apache.cassandra.locator.SimpleStrategy"). Could we find it elsewhere?

The reason I introduced this mapping is because I didn't find anything similar.

> But even more fundamentally, I don't see why this map is even needed: Since the name "SimpleStrategy" (for example) is the same one used on both CREATE statement and in the configuration, why can't we just compare names? Why do we need to go through translation to the locator::replication_strategy_type enum?

We could just operate on strings, would be easier, but then I thought we should somehow connect the string provided by a user in `scylla.yaml` to an actual type we use in the source code in order to avoid some weird mistakes and make maintenance easier (if someone ever removes a replication strategy, this map will have to be updated as well -> this place won't be missed by accident). 

Don't know if that reasoning is enough, though, seemed more "safe" to me.

Reply:
But who said that the name the user uses in scylla.yaml needs to be "network_topology"? Why can't it be the same name that the user already uses in CQL, which is NetworkTopologyStrategy? In fact, I wonder if it isn't more natural to list exactly the same names that is used in CQL and not some variant of them?

Regarding a map, we already have in locator/network_topology_strategy.cc the last two lines:
```
static registry registrator("org.apache.cassandra.locator.NetworkTopologyStrategy");
static registry registrator_short_name("NetworkTopologyStrategy");
```
Supposedly this adds to the "registry" these names, so their list is available somewhere (and probably used in the CreateKeyspace statement).  I don't know the details.

Your approach will also work, of course, but the downside is that if we ever invent new strategies (let's assume we do, because your choice of making this guardrail take a list of them and not be three separate guardrails), we'll need to remember to add them to yet more places (I think at least two new places in this code) or things will mysteriously break. For example, imagine that next year somebody will add TabletTopologyStrategy or CoolNewStrategy, and forget to add it to your new guardrail list, and nobody will notice anything is missing - until a year later somebody tries to warn about this strategy, and discover it is missing in the guardrail's list.

Reply:
That's a good idea and I used it. The only caveat was that, in order to get replication strategy's type, I have to first create a temp replication strategy object and only then get its type. But I guess we can live with that.